### PR TITLE
Redux

### PR DESCRIPTION
**Pull request by Dekim12**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,30 +1,34 @@
...
-import MainPage from './pages/MainPage/MainPage';
-import { HOME, FAVORITES, BEER_PAGE } from './constants';
-import './sass/App.sass';
+import { Header } from "./commonComponents";
+import BeerPage from "./pages/BeerPage/BeerPage";
+import FavoritesPage from "./pages/FavoritesPage/FavoritesPage";
+import MainPage from "./pages/MainPage/MainPage";
+import { HOME, FAVORITES, BEER_PAGE } from "./constants";
+import "./sass/App.sass";
+import { Provider } from "react-redux";
```
Comment (Dekim12) [inline src/App.jsx]: Please, review all imports

File: src/index.js [hunk 2]
```diff
@@ -3,12 +3,16 @@ import ReactDOM from 'react-dom';
...
 import App from './App';
+import { Provider } from 'react-redux';
+import store from './redux/store/store';
 
 ReactDOM.render(
   <React.StrictMode>
-    <BrowserRouter>
-      <App />
-    </BrowserRouter>
+    <Provider store={store}>
```
Comment (Dekim12) [inline src/index.js:11]: Why do you use redux provider twice? In the index and App files

File: src/pages/BeerPage/BeerPage.jsx [hunk 3]
```diff
@@ -1,101 +1,56 @@
...
+import { useLocation } from "react-router-dom";
 
-import { Button } from '../../commonComponents';
-import './style.sass';
+import BeerHead from "./Components/BeerHead/BeerHead";
+import BeerSpecification from "./Components/BeerSpecification/BeerSpecification";
+import BeerBrewing from "./Components/BeerBrewing/BeerBrewing";
+import BeerIngredients from "./Components/BeerIngredients/BeerIngredients";
+import BeerMethod from "./Components/BeerMethod/BeerMethod";
+import "./style.sass";
```
Comment (Dekim12) [inline src/pages/BeerPage/BeerPage.jsx]: move stales import to the end

File: src/pages/BeerPage/BeerPage.jsx [hunk 4]
```diff
@@ -1,101 +1,56 @@
...
-            <tr>
-              <h3>Yeast</h3>
-              <p>25 litres</p>
-            </tr>
-          </table>
-        </div>
+function BeerPage() {
+  const location = useLocation();
+  const dispatch = useDispatch();
+  let id = location.pathname.split("/"); //do this immutable
```
Comment (Dekim12) [inline src/pages/BeerPage/BeerPage.jsx]: [1,2,3].slice(-1)[0]

File: src/pages/BeerPage/BeerPage.jsx [hunk 5]
```diff
@@ -1,101 +1,56 @@
...
+  id = id[id.length - 1];
+
+  const selectBeerById = (state, id) => {
+    return state.BeerReducer.beerData.find((obj) => obj.id === +id);
```
Comment (Dekim12) [inline src/pages/BeerPage/BeerPage.jsx]: better to use redux selectors to get any data from redux-store

File: src/pages/BeerPage/BeerPage.jsx [hunk 6]
```diff
@@ -1,101 +1,56 @@
...
+  };
+
+  useEffect(
+    () => dispatch(fetchAllData("https://api.punkapi.com/v2/beers")),
+    []
+  );
+
+  const beer = useSelector((state) => selectBeerById(state, id));
```
Comment (Dekim12) [inline src/pages/BeerPage/BeerPage.jsx]: define all necessary selectors and use them something like useSelector(selectBeerById)

File: src/pages/BeerPage/BeerPage.jsx [hunk 7]
```diff
@@ -1,101 +1,56 @@
...
+  const dispatch = useDispatch();
+  let id = location.pathname.split("/"); //do this immutable
+  id = id[id.length - 1];
+
+  const selectBeerById = (state, id) => {
+    return state.BeerReducer.beerData.find((obj) => obj.id === +id);
+  };
+
+  useEffect(
+    () => dispatch(fetchAllData("https://api.punkapi.com/v2/beers")),
```
Comment (Dekim12) [inline src/pages/BeerPage/BeerPage.jsx]: Save it into constant "https://api.punkapi.com/v2/beers". You should define it in the thunk not pass as parameter

File: src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+      <h3>Hops</h3>
+      <p>
+        {ingredients.hops.map((hopsItem) => {
+          return (
+            <p key={ingredients.hops.indexOf(hopsItem)}>
+              {'"'}
+              {hopsItem.name}
+              {'"'} - {hopsItem.amount.value} {hopsItem.amount.unit}, add when{" "}
+              {hopsItem.add}
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx]: is better to use the template string

File: src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+function BeerIngredients({ ingredients }) {
+  return (
+    <div className="ingredients">
+      <h2>Ingredients</h2>
+
+      <h3>Water</h3>
+      <p>25 litres</p>
+
+      <h3>Malt</h3>
+      {ingredients.malt.map((maltItem) => {
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx]: Please, define mapping outside of return block

File: src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        return (
+          <p key={ingredients.malt.indexOf(maltItem)}>
+            {maltItem.name} - {maltItem.amount.value} {maltItem.amount.unit}
+          </p>
+        );
+      })}
+
+      <h3>Hops</h3>
+      <p>
+        {ingredients.hops.map((hopsItem) => {
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx]: the same

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+import "./style.sass";
+
+function BeerMethod({ method }) {
+  return (
+    <div className="method">
+      <h2>Method</h2>
+
+      <h3>Mash</h3>
+      {method.mash_temp.map((item) => {
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: Is better to keep it outside return block

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        return (
+          <p key={method.mash_temp.indexOf(item)}>
+            {item.duration || 0} minutes at {item.temp.value}{" "}
+            {item.temp.unit === "celsius" ? "℃" : "℉"}{" "}
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: use template string

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+          </p>
+        );
+      })}
+
+      <h3>Fermentation</h3>
+      <p>
+        Perform at {method.fermentation.temp.value}
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: use template string

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        {method.fermentation.temp.unit === "celsius" ? "℃" : "℉"}
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: create the utils function for defining measurable values

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+      </p>
+
+      {method.twist ? (
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: method.twist &&  something is enough

File: src/pages/BeerPage/Components/BeerSpecification/BeerSpecification.jsx [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+import BeerProperties from "../BeerProperties/BeerProperties";
+import BeerFoodPairing from "../BeerFoodPairing/BeerFoodPairing";
+import "./style.sass";
+
+function BeerSpecification({ ABVValue, IBUValue, EBCValue, foodPairing }) {
+  return (
+    <div className="beer-specification">
+      <BeerProperties
+        ABVValue={ABVValue}
```
Comment (Dekim12) [inline src/pages/BeerPage/Components/BeerSpecification/BeerSpecification.jsx]: don't use props names from capital letter

File: src/pages/MainPage/MainPage.jsx [hunk 17]
```diff
@@ -1,25 +1,26 @@
...
+import { Loading } from "../../commonComponents";
+import Search from "./Components/Search/Search";
+import { BEER_PAGE } from "../../constants";
+import "./style.sass";
 
 function MainPage() {
-  const [data, setData] = useState(null);
+  const dispatch = useDispatch();
+  useEffect(
+    () => dispatch(fetchAllData("https://api.punkapi.com/v2/beers")),
```
Comment (Dekim12) [inline src/pages/MainPage/MainPage.jsx]: the same as mentioned before

File: src/pages/MainPage/MainPage.jsx [hunk 18]
```diff
@@ -1,25 +1,26 @@
...
+    []
+  );
 
-  useEffect(async () => {
-    const response = await fetch('https://api.punkapi.com/v2/beers');
-    const data = await response.json();
-    setData(data);
-  }, []);
+  let beerCardsCollection = useSelector((state) => state.BeerReducer.beerData);
```
Comment (Dekim12) [inline src/pages/MainPage/MainPage.jsx]: use specific selector

File: src/redux/beerDataSlice.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+/* eslint-disable */
+export const fetchAllData = createAsyncThunk(
```
Comment (Dekim12) [inline src/redux/beerDataSlice.js]: Keep thunk middlewares in one place outside of redux slices

File: src/index.js (hunk 2 above)
Comment (Finnchick) [reply src/index.js:11]: Done

File: src/pages/BeerPage/BeerPage.jsx (hunk 7 above)
Comment (Finnchick) [reply src/pages/BeerPage/BeerPage.jsx]: Done

File: src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx (hunk 8 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx]: Done

File: src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx (hunk 9 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx]: Done

File: src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx (hunk 10 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerIngredients/BeerIngredients.jsx]: Done

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx (hunk 15 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: Done

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx (hunk 14 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: Done

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx (hunk 13 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: Done

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx (hunk 12 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: Done

File: src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx (hunk 11 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerMethod/BeerMethod.jsx]: Done

File: src/pages/BeerPage/BeerPage.jsx (hunk 4 above)
Comment (Finnchick) [reply src/pages/BeerPage/BeerPage.jsx]: Done

File: src/pages/BeerPage/Components/BeerSpecification/BeerSpecification.jsx (hunk 16 above)
Comment (Finnchick) [reply src/pages/BeerPage/Components/BeerSpecification/BeerSpecification.jsx]: Done

File: src/redux/beerDataSlice.js (hunk 19 above)
Comment (Finnchick) [reply src/redux/beerDataSlice.js]: Done

File: src/pages/MainPage/MainPage.jsx (hunk 17 above)
Comment (Finnchick) [reply src/pages/MainPage/MainPage.jsx]: Done

File: src/pages/BeerPage/BeerPage.jsx (hunk 6 above)
Comment (Finnchick) [reply src/pages/BeerPage/BeerPage.jsx]: Done (special thx)

File: src/pages/BeerPage/BeerPage.jsx (hunk 5 above)
Comment (Finnchick) [reply src/pages/BeerPage/BeerPage.jsx]: Done

File: src/pages/MainPage/MainPage.jsx (hunk 18 above)
Comment (Finnchick) [reply src/pages/MainPage/MainPage.jsx]: Will be done in near future :)

**Pull request closed without merging**